02_knight_and_princess/tevfik_aksoy/main.cpp

```diff
@@ -413,106 +275,81 @@ void generate_green_edges(Graph& graph) {
+    for (VertexId j = 0; j < vertices_in_depth.size() - 1; j++) {
+      const auto source = vertices_in_depth[j];
+      const auto destination = vertices_in_depth[j + 1];
       if (probability(condition)) {
-        graph.insert_edge(i, i + 1, Edge::Color::Blue);
+        graph.insert_edge(source, destination, Edge::Color::Blue);
       }
+    }
   }
 }
 
-int created_by_vertex(const Graph& graph, const VertexId& id);
-
-VertexId get_random_vertex_for_yellow(const Graph& graph,
-                                      const VertexDepth& depth,
-                                      const VertexId& source);
-
+vector<VertexId> filter_unconnected(const VertexId& vertex,
+                                    const vector<VertexId>& vertices_to_filter,
+                                    const Graph& graph) {
+  vector<VertexId> filtered_vertices;
+  for (const auto& v : vertices_to_filter) {
+    if (graph.get_source_of_vertex(v) != vertex) {
+      filtered_vertices.push_back(v);
+    }
+  }
+  return filtered_vertices;
+}
 void generate_yellow_edges(Graph& graph) {
-  int m = 0, n = 0;
-  auto max_depth = graph.get_vertex_depth(graph.get_vertices().size() - 1);
-
-  for (VertexDepth depth = 1; depth < max_depth; depth++) {
-    float condition = 100 - depth * (100 / (float)(max_depth - 1));
-    while (m < graph.vertices_count_in_depth(depth)) {
-      while (n < graph.vertices_count_in_depth(depth + 1)) {
-        if (probability(condition)) {
-          VertexId source = graph.vertices_in_depth(depth).at(m);
-          VertexId destination =
-              get_random_vertex_for_yellow(graph, depth, source);
-
-          if (destination == INVALID_ID)
+  for (VertexDepth depth = 0; depth < graph.depth(); depth++) {
+    float condition = 100 - depth * (100 / (float)(graph.depth() - 1));
+    const auto& vertices = graph.get_vertices_in_depth(depth);
+    const auto& vertices_next = graph.get_vertices_in_depth(depth + 1);
+    for (const auto& vertex : vertices) {
+      if (probability(condition)) {
+        const auto& filtered_next_vertices =
+            filter_unconnected(vertex, vertices_next, graph);
+        VertexId next_random_vertex_id;
+        bool break_flag = false;
+        do {
+          if (filtered_next_vertices.size() == 0) {
+            // All the vertices in the next depth are created by this
+            // vertex, so we break out of the infinite loop
+            break_flag = true;
             break;
-
-          graph.insert_edge(source, destination, Edge::Color::Yellow);
-          break;
-        }
-        n++;
+          }
+          next_random_vertex_id = get_random_vertex_id(filtered_next_vertices);
+        } while (graph.get_source_of_vertex(next_random_vertex_id) == vertex);
+        if (!break_flag)
+          // if we've broken out, then there is no random vertex generated
+          graph.insert_edge(vertex, next_random_vertex_id, Edge::Color::Yellow);

```

Comment:
Что ты здесь такое делаешь заумное?
```suggestion
        if (filtered_next_vertex_ids.size() == 0) {
          continue;
        }
        VertexId random_vertex_id = get_random_vertex_id(filtered_next_vertex_ids);
        graph.insert_edge(vertex_id, random_vertex_id, Edge::Color::Yellow);
```
Этого недостаточно?

Reply:
Кажется да, достаточно 😅

Reply:
Не нажимай `Resolved`, если вместо исправления оставил комментарий, который продолжает дискуссию.

А я тебе говорю достаточно. Ты передаешь фильтрованный вектор, и тебе нужно вернуть один элемент из него. Эта вся логика, которая тебе нужна. Не вижу никакой проблемы.

Reply:
+